packages/celotool/src/cmds/bots/auto-vote.ts

```diff
@@ -0,0 +1,204 @@
+  const election = await kit.contracts.getElection()
+  const validators = await kit.contracts.getValidators()
+
+  const groupScores = new Map<string, BigNumber[]>()
+  for (const valSigner of await election.getCurrentValidatorSigners()) {
+    try {
+      const val = await validators.getValidatorFromSigner(valSigner)
+      const groupAddress = val.affiliation!
+      if (groupScores.has(groupAddress)) {
+        groupScores.get(groupAddress)!.push(val.score)
+      } else {
+        groupScores.set(groupAddress, [val.score])
+      }
+    } catch (error) {
+      console.info(`could not get info for ${valSigner}, skipping...`)
+    }
+  }
+
+  const groupWeights = new Map<string, BigNumber>()
+
+  // Add up the scores for each group's validators and calculate weights
+  for (const group of groupScores.keys()) {
+    const scores = groupScores.get(group)!
+    // This shouldn't end up being 0, but handle the edge case and avoid division by 0
+    if (scores.length > 0) {
+      const avg = scores.reduce((a, b) => a.plus(b), new BigNumber(0)).div(scores.length)
+      groupWeights.set(group, avg.pow(20))
+    } else {
+      groupWeights.set(group, new BigNumber(0))
+    }
+  }
+
+  return groupWeights
+}
+
+function getWeightedRandomChoice(
+  groupWeights: Map<string, BigNumber>,
+  groupsToConsider: string[]
+): string {
+  // Filter to groups open to consideration, and sort from highest probability to lowest
+  const sortedGroupKeys = [...groupWeights.keys()]
+    .filter((k) => groupsToConsider.includes(k))
+    .sort((a, b) => {
+      return groupWeights.get(b)!.comparedTo(groupWeights.get(a)!)
+    })
+
+  let weightTotal = new BigNumber(0)
+  for (const key of sortedGroupKeys) {
+    weightTotal = weightTotal.plus(groupWeights.get(key) || 0)
+  }
+
+  const choice = weightTotal.multipliedBy(Math.random())
+  let totalSoFar = new BigNumber(0)
+
+  for (const key of sortedGroupKeys) {
+    totalSoFar = totalSoFar.plus(groupWeights.get(key)!)
+    if (totalSoFar.isGreaterThanOrEqualTo(choice)) {
+      return key
+    }
+  }
+  // This shouldn't happen, since the accumulated total at the last key of the
+  // sorted groups is always going to be higher than the value for `choice`.
+  // This is here to handle unknown edge-cases and to make Typescript happy
+  throw Error('An unexpected error occured when choosing a group via weighted random choice')
+}
+
+async function activatePendingVotes(kit: ContractKit, botKeys: string[]): Promise<void> {
+  const election = await kit.contracts.getElection()
+
+  for (const key of botKeys) {
+    kit.addAccount(key)
+    const account = ensure0x(getAccountAddressFromPrivateKey(key))
+    if (!(await election.hasActivatablePendingVotes(account))) {
+      try {
+        console.info(`activating votes for ${account}`)
+        const activateTxs = await election.activate(account)
+        for (const tx of activateTxs) {
```

Comment:
Consider that this will execute all txs serially which might not be desired. What do you think about doing them reasonably concurrently?